### ISSUE TITLE
Avoid latency spikes in the loading_cache class

### ISSUE THREAD

**Issue by duarten** [scylladb/scylladb#3693]
The `loading_cache` class can optionally reload values after a specified time period has elapsed. A timer runs at that frequency, and on each tick we reload all expired entries in parallel. This has several disadvantages:

- Requires allocating a data structure to hold the entries we decided to reload;
- It's possible that on each tick we reload a large amount of entries, resulting in latency spikes;
- Doing the reload in parallel schedules a lot of continuations in one go, which requires a lot of allocations and can trigger our max task backlog threshold, above which we ignore I/O;
- For some types of the cache, reloading an item doesn't defer and it can cause us to consume the whole task quota.

A way to solve this is to randomize the refresh time for each item, taking the configured period as the reference. We can have a list sorting the entries by refresh time, or associate a timer with each of them.

**Comment by hakuch** [comment]
@duarten, I'd like to try to address this issue.

When you say

> taking the configured period as the reference

do you mean that a cache with a refresh period `t` will refresh each entry after time `t + r(i)` where `r(i)` is a random variable?

> We can have a list sorting the entries by refresh time, or associate a time with each of them.

So that the cache can refresh the entries in the correct order, yes?

**Comment by duarten** [comment]
Yes on both counts :)

**Comment by hakuch** [comment]
Great!

I wonder how much wiggle room in the refresh period the cache has so that it can still satisfy it's configured properties. Suppose we allow for 1% variation. If the refresh period is already quite small, then the random variation in the refresh period could be smaller than the granularity of the timer.

**Comment by tzach** [comment]
@duarten is this issue related to a regression in 2.3-rc1?

**Comment by duarten** [comment]
@tzach Nop.

**Comment by duarten** [comment]
You can look at what current callers specifies and enforce (`static_assert`) a lower bound. The refresh period should be at least an order of magnitude greater  than the timer resolution.

**Comment by vladzcloudius** [comment]
@duarten @hakuch Another way to resolve this is to simply "reload" one item at a time instead of doing it in parallel. The advantage of this approach is that it's a lot simpler.

Randomizing on the other hand has a disadvantage of still allowing the latency spilkes under certain conditions and most importantly it *requires* the separate timer/thread object per item which is a very high price. 

As far as I can tell the randomization doesn't have any advantage compared to the one-at-a-time reloading.

**Comment by duarten** [comment]
Our timers are cheap, and we can probably get away with scheduling just a singler timer. Doing it one at a time, but starting at the same instant, has the disadvantage that we'll be perturbing the normal workload for a potentially large stretch of time.

**Comment by vladzcloudius** [comment]
@duarten I didn't know our timers are cheap - according to @gleb-cloudius it's not the case. Unless something has changed since my work on hinted handoff.
As for preventing other tasks from working - we should use `yield()` in order to ensure the required task rescheduling granularity as it has been discussed in detail in "[PATCH scylla v1 1/3] utils::loading_cache::on_timer(): use chunked_vector for storing elements we want to reload" on the scylladb-dev list. As I mentioned on that thread a few times - randomizing won't resolve the main problem - avoiding latency spikes.

**Comment by duarten** [comment]
I don't think anything changed, but timers seem cheap to me. Anyway, we can get away with a single one. 

I also don't think any solution prevents latency spikes. You need to at least use CPU to do the reloading, and that's stealing CPU away from the current workload, consequently increasing its latency. Reloading all items at the refresh time, even if one by one, adds latency to user requests during a relatively high period of time. It's better to randomize within a delta of the refresh time, and also load items one by one, but not all of them. 

The issue with adding latency for a stretch of time is that it can cause queues to fill up and the system to enter a bad, over-utilized state, whereas by adding some randomization, you're allowing the system more breathing room.

**Comment by vladzcloudius** [comment]
If you reload items one-by-one and yield() after each item you should be minimizing the negative effect of extra CPU consumption you have mentioned. This is an extreme approach. To mild it up a little we may yield every X time during the reload. This is exactly what I proposed on the thread above - not just reload one-by-one. So, in my proposal there isn't any "added latency for a stretch of time".

Since we agree that we want to load items one-by-one let's see if randomizing is going to give us any extra value since it seems to be the only difference between my and your's proposals.

I claim that the randomization is a premature optimization and not the central part of your algorithm because with or without it you still need to implement all elements of my proposal:
  * yield()ing if reloading items takes "too long" (the definition of "too long" may differ)
  * load items one-by-one

...and I suggest to leave it to the later stage after we see that this is really needed.

If all you add is Just a randomization (without yield()ing) it is not going to help because you will still reload all items every time timer is going to fire: today timer fires every "reload" period and items are reloaded if they are "older" than this period. Therefore no matter how you randomize their TS inside this period they are still going to "age out" all together starting from the second timer iteration.

To overcome this you have to decrease the timer period. Since you were talking about the sorted list of items' reload events I assume you have a scheduling-at-the-next-reload-event-time-point idea in mind. 

First of all, maintaining a sorted list of thousands of items on each shard is going to be quite expensive. And inserting new items (O(log(N)) complexity) may not be a rare event if we hit the maximum cache size limitation.

Second, if loading a single item is taking a long time you can still find yourself reloading a lot of items in the same timer iteration. 

Today `loading_cache_clock_type` is `lowres_clock`. Unless we want to change that this may also become a reason for items "coalescing" in the same timer iteration.

That's why you'll still need to yield() in order to avoid the "adding latency for a stretch of time".

Therefore the difference between the randomized and my proposal is a more frequent timer and the overhead of sorting the list of timestamps. The later may become a very serious source of latency by itself.

But let's consider the best case scenario for your algorithm - let's assume you've succeeded to achieve your goal of distributing all items nicely across the "reload" time frame, and they never coalesce, and items are never evicted from the cache (thus you don't need to frequently insert items into the sorted list). Is it going to make the system behave any better than in my proposal where you yield() after each reloaded item? And I think that obviously the answer is "no". And if we yield() after each X time (e.g. 50us) my proposal is going to do even better job since it will have less rescheduling events and less timer scheduling events than in your suggestion while still keeping the added latency under the sane threshold, and it's going to complete the "reloading" a lot faster minimizing the periods of "added latency due to reloading" all together.

However I think that "yield() every X time" may fall under the same category as randomizing - it's a preliminary optimization. I think simply yield()ing after each item will be good enough for the first phase.

**Comment by glommer** [comment]
About timers: as far as I understand, expiring lowres_clock based timers should be cheap as they don't involve signals and are just expired from the pollers.

The high res timers are the expensive ones.

But aside from that, from what I could understand from the discussion I think Vlad is right. If some elements alone can burst the task quota we have to yield after each anyway so the simple thing would be to just go over each of them in a simple list. We just need to make sure they are properly tagged in the main group.

**Comment by duarten** [comment]
@vladzcloudius 

> If you reload items one-by-one and yield() after each item you should be minimizing the negative effect of extra CPU consumption you have mentioned. This is an extreme approach. To mild it up a little we may yield every X time during the reload. This is exactly what I proposed on the thread above - not just reload one-by-one. So, in my proposal there isn't any "added latency for a stretch of time".

That doesn't minimize the negative effect of CPU consumption. In fact, it aggravates it. If you yield only X items (how do you calculate X?), you can easily blow the task quota.

> I claim that the randomization is a premature optimization and not the central part of your algorithm because with or without it you still need to implement all elements of my proposal:
> 
> yield()ing if reloading items takes "too long" (the definition of "too long" may differ)
> load items one-by-one

I would use a `do_for_each`, which checks for preemption in case the reload function is synchronous, and collect the unresolved futures returned by the reload function, thus reloading items in parallel if they need to do IO. 

> First of all, maintaining a sorted list of thousands of items on each shard is going to be quite expensive. And inserting new items (O(log(N)) complexity) may not be a rare event if we hit the maximum cache size limitation.

I don't see why. The timer wheel uses buckets, so it amortizes a lot of the complexity. If we keep a simple queue Q in the loading_cache when we expire X items upon a timer firing, then we assign new expiration times to each of the X items in the interval [max(Q.back().expiration + max_delta, now() + refresh_delay). This way we just sort the X items and push them back to Q.

> Second, if loading a single item is taking a long time you can still find yourself reloading a lot of items in the same timer iteration.

No, because you would gather the items that need to be reloaded before doing anything else, and you would only load those items. Loading items shouldn't take too long, and if it does, we degenerate into an approach similar to yours.

>That's why you'll still need to yield() in order to avoid the "adding latency for a stretch of time".

Yielding doesn't avoid that, since you'll be perturbing the workload for longer, increasing chances of saturation. Making the items cheap to reload and spreading their reload across a time period alleviates this.

>Therefore the difference between the randomized and my proposal is a more frequent timer and the overhead of sorting the list of timestamps. The later may become a very serious source of latency by itself.

The difference is not that, the difference is that is doing it x-by-x can bust the task quota, and perturbes the system with a potentially long-running background operation.

> Is it going to make the system behave any better than in my proposal where you yield() after each reloaded item? And I think that obviously the answer is "no". And if we yield() after each X time (e.g. 50us) my proposal is going to do even better job since it will have less rescheduling events and less timer scheduling events than in your suggestion while still keeping the added latency under the sane threshold, and it's going to complete the "reloading" a lot faster minimizing the periods of "added latency due to reloading" all together.

I don't agree. Yielding after X times has higher overhead than even scheduling a timer, since you have to schedule the thread out and then back in. I also disagree that it keeps latency under the same bounded threshold. And finally, the goal is not to reload faster, it's to spread reload operations in time so the system doesn't have a background task running for too long, competing with other important workloads.

**Comment by duarten** [comment]
@glommer 
>But aside from that, from what I could understand from the discussion I think Vlad is right. If some elements alone can burst the task quota we have to yield after each anyway so the simple thing would be to just go over each of them in a simple list. We just need to make sure they are properly tagged in the main group.

Isn't that the worse thing you can do? Keep busting the task quota in a loop? That's a great reason to reload items across time IMO.

**Comment by glommer** [comment]
Without the CPU scheduler that's a big problem.
But with the CPU scheduler I don't see much of an issue. As long as we are respecting the task quota (and not blowing it up), only the group in which we are scheduling this will have latency issues.

But we shouldn't be scheduling this in any latency-sensitive group.

As of today, statements are already completely separate in their own group.

**Comment by avikivity** [comment]
What group will this be scheduled in?

**Comment by tgrabiec** [comment]
@glommer I think that having a CPU scheduler doesn't mean we can completely dismiss latency impact of this. Each such group can add 500us to every deferring point (which defers) on the request processing path.

**Comment by glommer** [comment]
@avikivity as far as I understand, it should be main.

But that's accounted for in the latency guarantees we give. If this isn't running something else will be in that group.

If this has to run, and it takes 500usec, the the difference between this and randomizing runs will be that we will have requests with added 500usec in a row, versus requests with 500usec spikes in random times.

**Comment by avikivity** [comment]
Anything in main adds to the latency of everything else. So even if we yield, that's not enough. We should spread the work around.

**Comment by vladzcloudius** [comment]
On 08/25/2018 03:45 AM, Duarte Nunes wrote:
>
> @vladzcloudius <https://github.com/vladzcloudius>
>
>     If you reload items one-by-one and yield() after each item you
>     should be minimizing the negative effect of extra CPU consumption
>     you have mentioned. This is an extreme approach. To mild it up a
>     little we may yield every X time during the reload. This is
>     exactly what I proposed on the thread above - not just reload
>     one-by-one. So, in my proposal there isn't any "added latency for
>     a stretch of time".
>
> That doesn't minimize the negative effect of CPU consumption. In fact, 
> it aggravates it.
>

How yielding after each re-loaded item is going to aggravate it?

> If you yield only X items (how do you calculate X?), you can easily 
> blow the task quota.
>

If you read my proposal on the the thread I quoted before you would see 
the answer for your question. "X" is not a static value and I suggested 
to yield() when re-loading takes more-or-equal than a task quota. 
However in my previous post I suggested to start with yield()ing after 
each re-loaded item. So, let's stick with that for now and optimize 
later when/if we need it.

>     I claim that the randomization is a premature optimization and not
>     the central part of your algorithm because with or without it you
>     still need to implement all elements of my proposal:
>
>     yield()ing if reloading items takes "too long" (the definition of
>     "too long" may differ)
>     load items one-by-one
>
> I would use a |do_for_each|, which checks for preemption in case the 
> reload function is synchronous, and collect the unresolved futures 
> returned by the reload function, thus reloading items in parallel if 
> they need to do IO.
>

If you allow parallelism then you don't solve the original issue which 
is not only about re-loading not re-scheduling for a long time but also 
about firing 4000 async continuations at the same time on a single shard 
which will make the regular requests compete with them for resources and 
as a result increase their latency.
All this is discussed in detail on the thread I mentioned.

>     First of all, maintaining a sorted list of thousands of items on
>     each shard is going to be quite expensive. And inserting new items
>     (O(log(N)) complexity) may not be a rare event if we hit the
>     maximum cache size limitation.
>
> I don't see why. The timer wheel uses buckets, so it amortizes a lot 
> of the complexity. If we keep a simple queue Q in the loading_cache 
> when we expire X items upon a timer firing, then we assign new 
> expiration times to each of the X items in the interval 
> [max(Q.back().expiration + max_delta, now() + refresh_delay). This way 
> we just sort the X items and push them back to Q.
>

What's "timer wheel"?
Anyway you seem to miss my point completely. You don't need to sort 
anything when you handle the items that are already in the cache - you 
simply push them at the end of the list. However if your cache size is 
smaller than the number of items you need to store you will constantly 
evict items from the cache and insert new items into the cache - this is 
when you'll need to insert a new item into the sorted list and buckets 
or not it's still going to be a O(log(N)) operation.

>     Second, if loading a single item is taking a long time you can
>     still find yourself reloading a lot of items in the same timer
>     iteration.
>
> No, because you would gather the items that need to be reloaded before 
> doing anything else, and you would only load those items. Loading 
> items shouldn't take too long, and if it does, we degenerate into an 
> approach similar to yours.
>

And how are you going to implement this "degeneration"? IMO yield() is 
the simplest and the most obvious way to do it.

>     That's why you'll still need to yield() in order to avoid the
>     "adding latency for a stretch of time".
>
> Yielding doesn't avoid that, since you'll be perturbing the workload 
> for longer, increasing chances of saturation. Making the items cheap 
> to reload and spreading their reload across a time period alleviates this.
>

yielding (after each item re-loading) does the best job to avoid that 
because it is going to let the regular flow to run by imposing the least 
overhead possible.

>     Therefore the difference between the randomized and my proposal is
>     a more frequent timer and the overhead of sorting the list of
>     timestamps. The later may become a very serious source of latency
>     by itself.
>
> The difference is not that, the difference is that is doing it x-by-x 
> can bust the task quota, and perturbes the system with a potentially 
> long-running background operation.
>

After what I wrote above you should be able to understand that there 
isn't X-by-X in my proposal - it's 1-by-1.
"re-loading X items" was mentioned as an optional optimization and X is 
always calculated based on the time it took to reload these X elements 
in each particular timer iteration. In short: X is a number of items 
which time to reload is less-or-equal a particular time period T. I 
suggested T to be a task quota but Avi said it's not good enough. So, 
like I mentioned above, let's stick with yield() after each item. Do you 
still think my proposal has a higher chance to bust a task quota than 
yours? I think it's obvious that the answer is: "No, it's the other way 
around."

**Comment by vladzcloudius** [comment]
@tgrabiec That's exactly why we want to yield() after each reloaded item - if there are other (main flow) items to run they will get a run time.

@avikivity How randomizing is going to help anything compared to yield()ing after each item? Won't yield() let other work items run including from other scheduling groups?

**Comment by vladzcloudius** [comment]
We had a private talk with @duarten today. While I still think that randomization may be a preliminary optimization Duarte convinced me that it may be implemented in a very cheap and laconic way. So, I remove my objections to it.

**Comment by kostja** [comment]
@nuivall @ptrsmrn another issue manifesting the call for a full cache strategy in permissions.

**Comment by nyh** [comment]
> @nuivall @ptrsmrn another issue manifesting the call for a full cache strategy in permissions.

@kostja if "permissions" come from a Raft-backed table, why do we need loading_cache at all? It seems to me like we just need to read the entire table into memory once on boot, and then raft will ensure we will get all the changes later (or, will be told we missed changes and need to re-read everything). I don't think we need the "loading_cache" and its specific implementation that this issue is complaining about.

**Comment by nuivall** [comment]
> > @nuivall @ptrsmrn another issue manifesting the call for a full cache strategy in permissions.
> 
> @kostja if "permissions" come from a Raft-backed table, why do we need loading_cache at all? It seems to me like we just need to read the entire table into memory once on boot, and then raft will ensure we will get all the changes later (or, will be told we missed changes and need to re-read everything). I don't think we need the "loading_cache" and its specific implementation that this issue is complaining about.

I'm not sure what exactly you're suggesting? To extend row cache with ability to hold the whole table?

**Comment by ptrsmrn** [comment]
@nuivall can we close this? Or do we still need to wait for authorization cache?

**Comment by nuivall** [comment]
We need to wait.

**Comment by vladzcloudius** [comment]
> [@nuivall](https://github.com/nuivall) can we close this? Or do we still need to wait for authorization cache?

What do you mean, @ptrsmrn ?
I don't see any changes in the corresponding code of `loading_cache`.
I don't remember what did Duarte told me (https://github.com/scylladb/scylladb/issues/3693#issuecomment-416672516) but I do still see why implementing "smart yielding" using `need_preempt()` would make a lot of sense here (like we do at many other places in `scylla` code).

On top of that, we need to make sure we run these in the correct scheduling groups is `main` is not a good one (https://github.com/scylladb/scylladb/issues/3693#issuecomment-416255137).

So, @nuivall , could you, please, clarify what's the plan here?

**Comment by ptrsmrn** [comment]
@nuivall implemented https://github.com/scylladb/scylladb/pull/26841 and is working on authorization cache, which will close this issue.

**Comment by mykaul** [comment]
> [@nuivall](https://github.com/nuivall) implemented [#26841](https://github.com/scylladb/scylladb/pull/26841) and is working on authorization cache, which will close this issue.

@nuivall - the above is merged - what's left here?

**Comment by nuivall** [comment]
Replacing loading cache with auth cache (for authorization) is left to do.
